### DISCUSSION TITLE
Help understanding hydration for React 18

### DISCUSSION THREAD

**Discussion by litewarp** [RevereCRE/relay-nextjs#56] [General]
Hey there! Just wanted to say thanks for this library which has been working wonderfully for my projects.  

Now that React 18 has been released, I've been playing around with the [nextjs experimental runtimes](https://nextjs.org/docs/advanced-features/react-18/streaming) and have been trying to understand how to modify or implement a hydration strategy that allows use of the runtime. 

Right now, if you enable the `nodejs` experimental runtime, the client errs with a message saying that the preloaded query is null. Checking the server logs, it looks like the preloaded query is returned on the server when it's rendered for the stream, but not when the client tries to hydrate it with `usePreloadedQuery.`

What I can't unpack is whether this is a result of trying to hydrate the client twice -- once from `getServerSideProps` and then once again from the edge stream. I recall there being an issue a few months back, but would love to chat about it and/or help out implementing a streaming strategy for React 18.  

Thanks again!

**Comment by sibelius** [comment]
this is the example of Relay team with nextjs 

https://github.com/relayjs/relay-examples/pull/219

**Comment by rrdelaney** [comment]
As `relay-nextjs` encodes data into the document `<head>` I wouldn't expect it to work with streaming SSR at the moment. When Next.js gives guidance to the types of libraries that use this approach we'll have a path forward. This was addressed in a React 18 WG [discussion](https://github.com/reactwg/react-18/discussions/106), however Next.js doesn't expose that API for us to use and workaround.

**Comment by rrdelaney** [comment]
Also, the Relay team's example is awesome! Unfortunately it installs a global `getInitialProps` preventing static optimization of most pages. That doesn't work for Revere, so we still 100% plan on supporting this library.

**Comment by litewarp** [comment]
@rrdelaney Just wanted to chime in and say that the relay team [updated their example](https://github.com/relayjs/relay-examples/tree/main/data-driven-dependencies) to use `getServerSideProps`. 

Curious if you think this is substantially different that your current approach?

Thanks again!

**Reply by rrdelaney** [reply]
`getServerSideProps` is something intentionally avoided as in our experience the UX is much worse than `getInitialProps`. With `getInitialProps` we can start the navigation, show cached data, and refetch in the background. `getServerSideProps` necessarily performs a network roundtrip before navigation even appears to start and only finishes once all data is fetched. We actually started off with `getServerSideProps` but switched after user complaints.

**Reply by litewarp** [reply]
Interesting. Thanks for the explanation!
